src/dune_engine/action_exec.ml

```diff
@@ -138,6 +139,27 @@ module Exec_result = struct
   ;;
 end
 
+module Action_res = struct
+  type t =
+    { done_or_more_deps : Action_plugin.done_or_more_deps
+    ; needed_deps : Dep.Set.t
+    }
+
+  let union x y =
+    match x, y with
+    | ( { done_or_more_deps = x; needed_deps = a }
+      , { done_or_more_deps = y; needed_deps = b } ) ->
+      { done_or_more_deps = Action_plugin.done_or_more_deps_union x y
+      ; needed_deps = Dep.Set.union a b
+      }
+  ;;
+
+  let return ?needed_deps done_or_more_deps =
```

Comment:
Since you never pass `needed_deps` or `Need_more_deps _` at the same time, I suggest to have instead:
```
val no_deps: t
val need_more_deps: ... -> t
val actual_deps: Dep.Set.t -> t
```